lib/spack/spack/util/compression.py

```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
     return any(path.endswith(t) for t in ALLOWED_ARCHIVE_TYPES)
 
 
+def _tar(archive_file):
+    outfile = os.path.basename(archive_file)
+    remnant = os.path.join(os.getcwd(), outfile)
+    _, ext = os.path.splitext(archive_file)
+    try:
+        import tarfile
+        if ext in [".xz", ".txz"]:
```

Comment:
what happens for `.Z` (7zip) files?  I think they fall through this and Python tries and failed to use `tarfile` on them, then doesn't fall back to using `tar`.  Is that right?  Can you test this with `spack stage gblocks` (which I think is the only thing with a 7zip URL right now)?

Reply:
I completely forgot about 7zip, I'll add a catch for that. At the moment, you're right, currently it tries and fails with Python's native tarfile. In that vein, do we also want to add conditional support for the .7z extension now that Windows support is moving along?

Reply:
I looked into this a bit more and I think I'm wrong about what `.tar.Z` is.  It's actually not 7zip.  I think I was confused because I looked it up and the recommended program to use to decompress it on windows is 7zip.  See these:

* `.tar.Z` files: https://kb.iu.edu/d/acsy 
* `.Z` extension: https://kb.iu.edu/d/abck

The file signature is 1f 9d 90 -- which [according to wikipedia](https://en.wikipedia.org/wiki/List_of_file_signatures) is LZW compressed.

So yes we should have a special case here for `.Z` (and `.z` which I guess is also zipped).  I'm curious now whether they can be handled by either `zipfile` or `gzip` or some other core python module... but I couldn't really find anything.  So I you can't find anything, then yes I would fail this over to system tar (which seems to handle it fine).

On 7zip itself: yeah we should probably support it, but let's not do it in this PR since nothing currently uses it.  I am not sue what packages we would want to have 7zip downloads for, since it doesn't look like there is a common, builtin 7zip decompressor on linux or macOS.